### PR TITLE
Hackerforms postgres

### PR DESCRIPTION
**Pull request by lucasmachadorj**
# Hackerforms Postgres

A Client API for connecting with Postgres Databases and provide ready-to-use pages for CRUD use cases.

#### 1. Creating a Postgres Client Instance

```
>>> import hackerforms.crud.postgres as hpc
>>> db = hpc.Client(database=os.getenv('database'),user=os.getenv('user'),password=os.getenv('password'),host=os.getenv('host'))
```

##### Example: Create a insert form

Suppose a database in which there are three tables `products`, `suppliers` and `categories`. The table `products` has a reference to the tables `suppliers` and `categories` through the foreign key columns `category_id` and `category_id`, respectively.

In order to create a new product, we have to choose its category and supplier, so we are going to define two dropdowns for choosing them.

```
categories = db.dropdown(table="categories", column="category_name", primary_key='category_id') # [{'label': 'category_name', 'value': 'category_id'}]

```

```
suppliers = db.dropdown(table="suppliers", column="company_name", primary_key='supplier_id')
```

`suppliers` and `categories` has the dropdown options in the format expected by `read_dropdown` options. Its labels are the columns passed as parameter to the dropdown method and its values are the respective primary_keys.

Now, for creating the form:

```
page = (
    db.RowPage(table="products")
    .read_dropdown("category", options=categories, column="category_id")
    .read_dropdown("supplier", options=suppliers, column="supplier_id")
    .read("product name", key="product_name", column="product_name")
    .read_number("quantity per unit", column="quantity_per_unit")
    .read_currency("unit price", column="unit_price")
    .read_number("units in stock", column="units_in_stock")
    .read_number("units in order", column="units_on_order")
    .read_number("discontinued", column="discontinued")
    .insert(context={"product_id": 821, })
)
```

Notice that you have to provide the column linked to each widget. Besides that, we use the `suppliers` and `categories` variables as options to `read_dropdown` widgets. We finish our RowPage builder invoking the method insert(), which means it's a `create` form page. The parameter context is rather important. It's a dict with table column names as keys and its values are going to be persisted in row besides the values inserted in the form. Be aware that if you a column name in the form and in context object, the latter will override the value inserted in the form by the user.

##### Example: Search rows in a table

Now, imagine you wish to provide a page for your users search for rows in a table.
In our database example, we would like to allow the users to search for products by its name, display a table of results and update it dinamically while (s)he types, after all UX is crucial!

See below how to achieve it:

```
product = (
    db.SearchPage(table="products")
    .read("Product Name", column="product_name")
    .search(
        display_columns=["product_id", "product_name", "unit_price"]
    )
)
```

A page will be displayed with a `read` widget as input filter and a table will be updated while the user types, in realtime.

### CODE REVIEW

**Comment by lucasmachadorj** [comment]
image

File: hackerforms/crud/postgres/sql/queries.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+def get_column_values_query(table, *args):
+    columns = ""
+    for arg in args:
+        columns += f"{arg},"
+    columns = columns[:-1]
+    return f"select {columns} from {table}"
+
+
+def insert_new_row_query(table, page):
+    query = f"insert into {table}{tuple(page.keys())} "
```
Comment (bernardowerneck) [inline hackerforms/crud/postgres/sql/queries.py:10]: o valores de nome de tabela e colunas colocaria entre aspas duplas, pra n dar problemas

File: hackerforms/crud/postgres/sql/queries.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+def get_column_values_query(table, *args):
+    columns = ""
+    for arg in args:
+        columns += f"{arg},"
+    columns = columns[:-1]
+    return f"select {columns} from {table}"
```
Comment (bernardowerneck) [inline hackerforms/crud/postgres/sql/queries.py:6]: mesma coisa do insert

File: hackerforms/crud/crud_page.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    """
+
+    def __init__(self):
+        self.page = Page()
+        self.column_from_key = {}
+
+    def __validate_args(self, *args, **kwargs):
+        column = kwargs.get("column", None)
+        if not column or not args:
+            raise Exception
```
Comment (bernardowerneck) [inline hackerforms/crud/crud_page.py:26]: se colocar key="key", column="column" vai dar erro aqui n?

File: hackerforms/crud/postgres/postgres_search_page.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            if filters.get(key) != "" and key in self.column_from_key.keys()
+        ]
+        if len(filter_keys_with_nonempty_values) == 0:
+            return None
+        for i, key in enumerate(filter_keys_with_nonempty_values):
+            value = filters.get(key)
+            if i < len(filter_keys_with_nonempty_values) - 1:
+                query += f"{self.column_from_key[key]} like '%{value}%' and "
+            else:
+                query += f"{self.column_from_key[key]} like '%{value}%';"
```
Comment (bernardowerneck) [inline hackerforms/crud/postgres/postgres_search_page.py:47]: acho que deveria colocar " nas colunas

File: hackerforms/crud/crud_page.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+"""
+
+from abc import ABC
+from hackerforms.generated.page import Page
+
+
+class CRUDPage(ABC):
+    """Abstract class that serves as proxy to the Page Class widget methods.
+    Each proxy widget method sets the mapping between the database column name and the widget key.
+
```
Comment (bernardowerneck) [inline hackerforms/crud/crud_page.py:16]: acho que seria interessante ter os display na crud page tmb

**Pull request closed without merging**